mlir/lib/Target/LLVMIR/Dialect/OpenMP/OpenMPToLLVMIRTranslation.cpp

```diff
@@ -1638,13 +1631,154 @@ getRefPtrIfDeclareTarget(mlir::Value value,
   return nullptr;
 }
 
+// A small helper structure to contain data gathered
+// for map lowering and coalese it into one area and
+// avoiding extra computations such as searches in the
+// llvm module for lowered mapped varibles or checking
+// if something is declare target (and retrieving the
+// value) more than neccessary.
+struct MapData : llvm::OpenMPIRBuilder::MapInfosTy {
+  llvm::SmallVector<bool, 4> IsDeclareTarget;
+  llvm::SmallVector<mlir::Operation *, 4> MapClause;
+  llvm::SmallVector<llvm::Value *, 4> OriginalValue;
```

Comment:
Looks to me like `OriginalValue` holds same value as `Pointers`. If so, can we remove this?

Reply:
It doesn't in a few cases (https://github.com/llvm/llvm-project/pull/68689/files#diff-2cbb5651f4570d81d55ac4198deda0f6f7341b2503479752ef2295da3774c586R2265), OriginalValue should never change, the idea is that it's the original unaltered value pointed to by the captured MLIR map_info. The BasePointer/Pointers can be reassigned to new instructions such as GEP offsets, declare target globals, etc. and align with the BasePointer/Pointers that will be assigned to the kernel argument structure

Reply:
Probably worth adding as a comment.

Reply:
@agozillon please add this as a comment.